Привет! Поздравляю с первым выполненным проектом! Ты проделал большую работу. Я проверю твою работу и напишу свои комментарии. Они, в зависимости от результат проверки, будут в </font><font color='green'>зеленом (когда все отлично),</font> <font color='brown'>коричневом (когда можно лучше)</font> или <font color='red'>красном(когда нужно переделать, чтобы я принял работу)</font>. Постарайся учесть эти комментарии в дальнейших проектах.</font>

## Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

<font color='brown'>Здесь помимо постановки задачи я был бы рад видеть твое краткое видение целей исследования и описание данных.</font><br><font color='blue'>Цель — создать исследование для построении модели кредитного скоринга. Описание данных чуть ниже

### Шаг 1. Откройте файл с данными и изучите общую информацию. 

In [1]:
import pandas as pd
df = pd.read_csv('/datasets/data.csv')
print(df.head(10))

print(df.info())
print()
print(df['children'].value_counts())
print()
print(df['family_status'].value_counts())
print()
print(df['debt'].value_counts())
print()
print(df['income_type'].value_counts())
df



   children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0    

children  days_employed  dob_years education  education_id  \
0             1   -8437.673028         42    высшее             0   
1             1   -4024.803754         36   среднее             1   
2             0   -5623.422610         33   Среднее             1   
3             3   -4124.747207         32   среднее             1   
4             0  340266.072047         53   среднее             1   
...         ...            ...        ...       ...           ...   
21520         1   -4529.316663         43   среднее             1   
21521         0  343937.404131         67   среднее             1   
21522         1   -2113.346888         38   среднее             1   
21523         3   -3112.481705         38   среднее             1   
21524         2   -1984.507589         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

        total_income                       purpose  
0      253875.639453                 покупка жилья  
1      112080.014102       приобретение автомобиля  
2      145885.952297                 покупка жилья  
3      267628.550329    дополнительное образование  
4      158616.077870               сыграть свадьбу  
...              ...                           ...  
21520  224791.862382             операции с жильем  
21521  155999.806512          сделка с автомобилем  
21522   89672.561153                  недвижимость  
21523  244093.050500  на покупку своего автомобиля  
21524   82047.418899         на покупку автомобиля  

[21525 rows x 12 columns]

<font color='green'>Пожалуйста, подправь дублирование подсчета полей по семейному статусу</font><br>
<font color='green'>Done!</font>

### Вывод

Теоретически нужны только три столбца — семейное положение — 0/1 (int64, все строки заполнены), 
количество детей — все строки заполнены, int64



children — количество детей в семье // предварительно норм столбец, все integer. Посмотрю нет ли экстремальных
days_employed — общий трудовой стаж в днях // Отрицательные сделаю по модулю, округлю до integer, значения выше реалистичного трудового стажа удалю
dob_years — возраст клиента в годах // проверить чтобы были целые положительные, реалистичные
education — уровень образования клиента // сделать lowercase, удалить пробелы, посмотреть какие уникальные значения
education_id — идентификатор уровня образования // видимо, сделать бинарным — есть/нет
family_status — семейное положение // приведу к lower, уберу пробелы выделю уникальные значения,  уберу в словарь
family_status_id — идентификатор семейного положения // оставлю в основной таблице этот столбец без текстовой расшифровки
gender — пол клиента // убрать пробелы, lowercase
income_type — тип занятости // сделать lowercase, удалить пробелы, посмотреть какие уникальные значения
debt — имел ли задолженность по возврату кредитов // проверить бинарность — есть/нет
total_income — ежемесячный доход // категоризовать
purpose — цель получения кредита // сделать lowercase, удалить пробелы, посмотреть какие уникальные значения


<font color='green'>Тут все отлично, данные считаны, исследования запланированы. Использование словаря - это очень хорошо!</font>

### Шаг 2. Предобработка данных

#### Обработка пропусков

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


###### Подозрительно одинаковые пропуски в total_income (где деньги?) и days_employed, выведем таблицу

In [3]:
str_objects = ['education', 'family_status', 'gender', 'income_type', 'purpose']

for objects in str_objects:
    df[objects] = df[objects].str.lower()

In [4]:
df.pivot_table(index=['income_type'], values =['total_income', 'days_employed'], aggfunc =['count', 'mean'])

count                        mean               
                days_employed total_income  days_employed   total_income
income_type                                                             
безработный                 2            2  366413.652744  131339.751676
в декрете                   1            1   -3296.759962   53829.130729
госслужащий              1312         1312   -3399.896902  170898.309923
компаньон                4577         4577   -2111.524398  202417.461462
пенсионер                3443         3443  365003.491245  137127.465690
предприниматель             1            1    -520.848083  499163.144947
сотрудник               10014        10014   -2326.499216  161380.260488
студент                     1            1    -578.751554   98201.625314

##### Безработных с доходом, одного студента, пару предпринимателей и декрет удалим

In [5]:
df = df[df.income_type != 'безработный']
df = df[df.income_type != 'предприниматель']
df = df[df.income_type != 'студент']
df = df[df.income_type != 'в декрете']

In [6]:
df['income_type'].value_counts()

сотрудник      11119
компаньон       5085
пенсионер       3856
госслужащий     1459
Name: income_type, dtype: int64

In [7]:
df.pivot_table(index=['income_type'], values =['total_income', 'days_employed'], aggfunc =['count', 'mean', 'median'])

count                        mean                 \
            days_employed total_income  days_employed   total_income   
income_type                                                            
госслужащий          1312         1312   -3399.896902  170898.309923   
компаньон            4577         4577   -2111.524398  202417.461462   
пенсионер            3443         3443  365003.491245  137127.465690   
сотрудник           10014        10014   -2326.499216  161380.260488   

                    median                 
             days_employed   total_income  
income_type                                
госслужащий   -2689.368353  150447.935283  
компаньон     -1547.382223  172357.950966  
пенсионер    365213.306266  118514.486412  
сотрудник     -1574.202821  142594.396847

###### Пенсионеров с невероятным доходом и стажем удалим

In [8]:
df = df[df.income_type != 'пенсионер']
df.pivot_table(index=['income_type'], values =['total_income', 'days_employed'], aggfunc =['count', 'mean', 'median'])

count                       mean                 \
            days_employed total_income days_employed   total_income   
income_type                                                           
госслужащий          1312         1312  -3399.896902  170898.309923   
компаньон            4577         4577  -2111.524398  202417.461462   
сотрудник           10014        10014  -2326.499216  161380.260488   

                   median                 
            days_employed   total_income  
income_type                               
госслужащий  -2689.368353  150447.935283  
компаньон    -1547.382223  172357.950966  
сотрудник    -1574.202821  142594.396847

<font color='green'>Отлично подмечено. Но нужно обязательно сообщить об этом коллегам, занимающимся выгрузкой, пусть ищут ошибку.</font>

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 17663 entries, 0 to 21524
Data columns (total 12 columns):
children            17663 non-null int64
days_employed       15903 non-null float64
dob_years           17663 non-null int64
education           17663 non-null object
education_id        17663 non-null int64
family_status       17663 non-null object
family_status_id    17663 non-null int64
gender              17663 non-null object
income_type         17663 non-null object
debt                17663 non-null int64
total_income        15903 non-null float64
purpose             17663 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 1.8+ MB


##### Получаем названия колонок

In [10]:
#print(df.columns)
columns = ['children', 'days_employed', 'dob_years', 'education', 'education_id',
       'family_status', 'family_status_id', 'gender', 'income_type', 'debt',
       'total_income', 'purpose']

###### Проверим, совпадают ли пропуски в трудовом стаже и доходе

In [11]:
#data_with_nan = df[df['days_employed'] = df['days_employed'].isna()]
#data_with_nan
print(df[df['days_employed'].isnull()].count())


children            1760
days_employed          0
dob_years           1760
education           1760
education_id        1760
family_status       1760
family_status_id    1760
gender              1760
income_type         1760
debt                1760
total_income           0
purpose             1760
dtype: int64


In [12]:
data_with_nan = df[df['days_employed'].isnull()]
data_with_nan.pivot_table(index=['income_type'], 
                          values =['children', 'dob_years', 'education', 'family_status', 'debt', 'purpose'], 
                          aggfunc =['count', 'mean', 'median'])

count                                                  \
            children  debt dob_years education family_status purpose   
income_type                                                            
госслужащий      147   147       147       147           147     147   
компаньон        508   508       508       508           508     508   
сотрудник       1105  1105      1105      1105          1105    1105   

                 mean                        median                 
             children      debt  dob_years children debt dob_years  
income_type                                                         
госслужащий  0.687075  0.047619  41.231293        0    0        40  
компаньон    0.614173  0.059055  40.086614        0    0        40  
сотрудник    0.651584  0.088688  40.031674        0    0        40

In [13]:
data_with_nan

children  days_employed  dob_years education  education_id  \
26            0            NaN         41   среднее             1   
41            0            NaN         50   среднее             1   
65            0            NaN         21   среднее             1   
72            1            NaN         32    высшее             0   
82            2            NaN         50    высшее             0   
...         ...            ...        ...       ...           ...   
21489         2            NaN         47   среднее             1   
21495         1            NaN         50   среднее             1   
21497         0            NaN         48    высшее             0   
21502         1            NaN         42   среднее             1   
21510         2            NaN         28   среднее             1   

               family_status  family_status_id gender  income_type  debt  \
26           женат / замужем                 0      m  госслужащий     0   
41           женат / замужем                 0      f  госслужащий     0   
65     не женат / не замужем                 4      m    компаньон     0   
72           женат / замужем                 0      m  госслужащий     0   
82           женат / замужем                 0      f    сотрудник     0   
...                      ...               ...    ...          ...   ...   
21489        женат / замужем                 0      m    компаньон     0   
21495       гражданский брак                 1      f    сотрудник     0   
21497        женат / замужем                 0      f    компаньон     0   
21502        женат / замужем                 0      f    сотрудник     0   
21510        женат / замужем                 0      f    сотрудник     0   

       total_income                                purpose  
26              NaN                            образование  
41              NaN       сделка с подержанным автомобилем  
65              NaN  операции с коммерческой недвижимостью  
72              NaN  операции с коммерческой недвижимостью  
82              NaN                                  жилье  
...             ...                                    ...  
21489           NaN                   сделка с автомобилем  
21495           NaN                                свадьба  
21497           NaN             строительство недвижимости  
21502           NaN       строительство жилой недвижимости  
21510           NaN                приобретение автомобиля  

[1760 rows x 12 columns]

In [14]:
data_with_nan.groupby(['income_type']).count()

children  days_employed  dob_years  education  education_id  \
income_type                                                                
госслужащий       147              0        147        147           147   
компаньон         508              0        508        508           508   
сотрудник        1105              0       1105       1105          1105   

             family_status  family_status_id  gender  debt  total_income  \
income_type                                                                
госслужащий            147               147     147   147             0   
компаньон              508               508     508   508             0   
сотрудник             1105              1105    1105  1105             0   

             purpose  
income_type           
госслужащий      147  
компаньон        508  
сотрудник       1105

Судя по таблицам, пропуски случайные — есть у всех трех категорий людей и у всех целей.

##### Округлим до положительных целых количество рабочих дней , заполним средним или медианой

In [15]:
df['days_employed'] = abs(round(pd.to_numeric(df['days_employed'], errors='coerce')))
days_employed_median = df['days_employed'].median()
df['days_employed'] = df['days_employed'].fillna(value=days_employed_median)



##### Теперь разберемся с пропусками в Total Income. Округлим значения до целых, заполним средним или медианой. Значения похожи, возьмем среднее


In [16]:
df['total_income'] = round(pd.to_numeric(df['total_income'], errors='coerce'))
total_income_mean = df['total_income'].mean()
df['total_income'] = df['total_income'].fillna(value=total_income_mean)

In [17]:
df.pivot_table(index=['income_type'], values =['total_income', 'days_employed'], aggfunc =['count', 'mean', 'median'])

count                       mean                 \
            days_employed total_income days_employed   total_income   
income_type                                                           
госслужащий          1459         1459   3221.581905  171208.431363   
компаньон            5085         5085   2063.412783  199576.139970   
сотрудник           11119        11119   2257.282579  162632.050091   

                   median                 
            days_employed   total_income  
income_type                               
госслужащий        2385.0  161331.000000  
компаньон          1630.0  173976.308558  
сотрудник          1630.0  151820.000000

#### Пришла пора разобраться с возрастом

In [18]:
df = df[df.dob_years > 18]
df = df[df.dob_years < 65]

df['dob_years'].value_counts()

35    616
40    601
41    600
34    600
38    589
42    584
33    579
39    567
31    558
36    550
29    545
30    540
44    537
37    532
48    515
32    507
43    504
28    502
27    489
45    483
49    476
47    467
46    461
50    448
26    406
52    378
51    365
25    357
53    339
54    316
56    278
55    266
24    263
23    254
58    220
57    211
22    181
59    168
61    113
21    111
60    107
62     89
63     59
64     57
20     51
19     14
Name: dob_years, dtype: int64

### Вывод

Убрал странных людей по одному / двое — предприниматель, безработный, студент, в декрете.
Убрал всех пенсионеров — нереальный доход и стаж.

"days_employed, total_income — количественные переменные. Пропуски в таких переменных заполняют характерными значениями"
Заменим медианой или средним. Поскольку медиана значительно отличается от среднего в первом случае — заменяем медианой. Во втором значения похожи, земенил средним.

Возможные причины появления пропусков в данных — нет данных по доходам

<font color='red'>Нас должно насторожить, что банк дает кредиты людям без известного стажа и дохода. Случайны наши пропуски или нет?
    Можем ли мы выдвинуть какую-то  гипотезу, видя, что количество пропусков в двух полях совпадает? Можем ли мы ее проверить? Можем ли мы что-то сообщить коллегам, занимающимся выгрузкой?</font>
<br><font color='blue'>Людей без дохода и с невероятным стажем убрал. По гипотезам пока ничего не придумал(  
<font color='red'>Пропуски, видимо, симметричны, то есть у тех у кого нет данных по стажу, нет и по доходу. Хорошо было бы убедиться в этом. После чего нужно посмотреть, случайны ли эти пропуски. (Может быть, пропуски есть только у определенной категории людей?)</font>
    
<br><font color='blue'>Убедился, что пропуски симметричны и случайны — есть у всех категорий людей по доходу и целям. Заполняю медианой или средним.  
    <font color='green'>Хорошо, это был важный этап работы.</font>
    

<font color='green'>Что касается технической части, то ты молодец. Определены, а также корректно и правильными методами устранены все явные пропуски.</font> <font color='green'>Только вот как насчет неявных? Где обещанная проверка на реалистичность возрастов?  ;) <font><br><font color='blue'>Возраст проверил, убрал детей и старше 65!  
    <font color='green'>Вообще-то я намекал на людей с возрастом 0, т.к. это явно ошибка, но в удалении всех детей есть своя логика.</font>

   ### Замена типа данных

##### Модуль чисел количества детей и проверка

In [19]:
df['children'] = abs(df['children'])
df['children'].value_counts()

0     10439
1      4556
2      2021
3       322
20       66
4        40
5         9
Name: children, dtype: int64

Тип данных в детях — целые числа, ок.

In [20]:
#Проверяем информацию об объектах
df['days_employed'] = abs(df['days_employed'].astype('int'))
df['total_income'] = df['total_income'].astype('int')

df.info()
for name in columns:
    #print(f'Нулевых значений столбца {name} {df[name].isnull().sum()}')
    print()
    print(df[name].value_counts())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 17453 entries, 0 to 21524
Data columns (total 12 columns):
children            17453 non-null int64
days_employed       17453 non-null int64
dob_years           17453 non-null int64
education           17453 non-null object
education_id        17453 non-null int64
family_status       17453 non-null object
family_status_id    17453 non-null int64
gender              17453 non-null object
income_type         17453 non-null object
debt                17453 non-null int64
total_income        17453 non-null int64
purpose             17453 non-null object
dtypes: int64(7), object(5)
memory usage: 1.7+ MB

0     10439
1      4556
2      2021
3       322
20       66
4        40
5         9
Name: children, dtype: int64

1630     1748
200        15
206        14
1482       14
220        14
         ... 
10160       1
6312        1
8087        1
6038        1
2049        1
Name: days_employed, Length: 5657, dtype: int64

35    616
40    601
41    

### Вывод

Все строковые значения — object, цифровые привел к integer

<font color='green'>Отлично, теперь с типами данных проблем нет.</font>

### Обработка дубликатов

In [21]:
str_objects = ['education', 'family_status', 'gender', 'income_type', 'purpose']

for objects in str_objects:
    df[objects] = df[objects].str.lower() # создали новый столбец "objects"? Поправил на переменнуую вместо строкового значения!
    
df.head()

children  days_employed  dob_years education  education_id  \
0         1           8438         42    высшее             0   
1         1           4025         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4125         32   среднее             1   
5         0            926         27    высшее             0   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      f   сотрудник     0        253876   
1   женат / замужем                 0      f   сотрудник     0        112080   
2   женат / замужем                 0      m   сотрудник     0        145886   
3   женат / замужем                 0      m   сотрудник     0        267629   
5  гражданский брак                 1      m   компаньон     0        255764   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
5               покупка жилья

In [22]:
dublicated_raws = df[df.duplicated()]
dublicated_raws.head(50)


children  days_employed  dob_years education  education_id  \
2849          0           1630         41   среднее             1   
4182          1           1630         34    высшее             0   
6312          0           1630         30   среднее             1   
9238          2           1630         34   среднее             1   
10697         0           1630         40   среднее             1   
11791         0           1630         47   среднее             1   
12373         0           1630         58   среднее             1   
12375         1           1630         37   среднее             1   
13025         1           1630         44   среднее             1   
13773         0           1630         35   среднее             1   
13878         1           1630         31   среднее             1   
13942         0           1630         44   среднее             1   
14097         0           1630         48   среднее             1   
14432         2           1630         36    высшее             0   
14728         0           1630         46   среднее             1   
14832         0           1630         50   среднее             1   
15091         0           1630         58   среднее             1   
15188         0           1630         60   среднее             1   
15991         0           1630         51   среднее             1   
16176         0           1630         47   среднее             1   
16378         0           1630         46   среднее             1   
16902         2           1630         39   среднее             1   
16904         1           1630         32    высшее             0   
17379         0           1630         54    высшее             0   
17755         1           1630         43   среднее             1   
17774         1           1630         40   среднее             1   
18328         0           1630         29    высшее             0   
18349         1           1630         30    высшее             0   
18521         0           1630         56   среднее             1   
18563         0           1630         54   среднее             1   
19041         0           1630         56   среднее             1   
19184         0           1630         46   среднее             1   
19295         0           1630         51   среднее             1   
19321         0           1630         23   среднее             1   
19369         0           1630         45   среднее             1   
19387         0           1630         38    высшее             0   
19559         0           1630         41   среднее             1   
19832         0           1630         48   среднее             1   
19946         0           1630         57   среднее             1   
20165         0           1630         42   среднее             1   
20297         1           1630         23   среднее             1   
20662         0           1630         58   среднее             1   
21132         0           1630         47   среднее             1   
21281         1           1630         30    высшее             0   

               family_status  family_status_id gender  income_type  debt  \
2849         женат / замужем                 0      f    сотрудник     0   
4182        гражданский брак                 1      f    сотрудник     0   
6312         женат / замужем                 0      m    сотрудник     0   
9238         женат / замужем                 0      f    сотрудник     0   
10697       гражданский брак                 1      f    компаньон     0   
11791        женат / замужем                 0      f    сотрудник     0   
12373        женат / замужем                 0      m    сотрудник     0   
12375        женат / замужем                 0      f    сотрудник     0   
13025        женат / замужем                 0      f    сотрудник     0   
13773       гражданский брак                 1      f    сотрудник     0   
13878        женат / замужем                 0      f    компаньо

In [23]:
df.duplicated().sum()

44

In [24]:
dublicated_raws.describe()

children  days_employed  dob_years  education_id  family_status_id  \
count  44.000000           44.0  44.000000     44.000000         44.000000   
mean    0.363636         1630.0  43.159091      0.818182          0.522727   
std     0.613450            0.0   9.875841      0.390154          0.901901   
min     0.000000         1630.0  23.000000      0.000000          0.000000   
25%     0.000000         1630.0  35.750000      1.000000          0.000000   
50%     0.000000         1630.0  44.000000      1.000000          0.000000   
75%     1.000000         1630.0  50.250000      1.000000          1.000000   
max     2.000000         1630.0  60.000000      1.000000          4.000000   

       debt  total_income  
count  44.0          44.0  
mean    0.0      173976.0  
std     0.0           0.0  
min     0.0      173976.0  
25%     0.0      173976.0  
50%     0.0      173976.0  
75%     0.0      173976.0  
max     0.0      173976.0

In [25]:
dublicated_raws.pivot_table(index=['income_type'], 
               columns =['total_income', 'days_employed'], 
               aggfunc =['count', 'mean'])

count                                                        \
              children   debt dob_years education education_id family_status   
total_income    173976 173976    173976    173976       173976        173976   
days_employed     1630   1630      1630      1630         1630          1630   
income_type                                                                    
госслужащий          2      2         2         2            2             2   
компаньон            7      7         7         7            7             7   
сотрудник           35     35        35        35           35            35   

                                                   mean                    \
              family_status_id gender purpose  children   debt  dob_years   
total_income            173976 173976  173976    173976 173976     173976   
days_employed             1630   1630    1630      1630   1630       1630   
income_type                                                                 
госслужащий                  2      2       2  1.500000      0  33.000000   
компаньон                    7      7       7  0.285714      0  42.714286   
сотрудник                   35     35      35  0.314286      0  43.828571   

                                             
              education_id family_status_id  
total_income        173976           173976  
days_employed         1630             1630  
income_type                                  
госслужащий       0.000000         0.000000  
компаньон         0.714286         0.714286  
сотрудник         0.885714         0.514286

In [26]:
df = df.drop_duplicates().reset_index(drop=True)
df.duplicated().sum()

0

<font color='red'>В сэте по-прежнему 71 дубликат. С этим нужно что-то сделать. (А также выдвинуть гипотезы об источнике дубликатов) </font><br><font color='blue'> Дубликаты убрал!  
    <font color='red'>Молодец, что переписал код, но нужно исправить ошибку и подумать над гипотезами, откуда могли возникнуть дубликаты. Это отдельный и важный пункт в списке заданий.</font>
    
<br><font color='blue'> Вероятно это случайные совпадения или повторения строк copy/paste или дублирующейся выгрузкой.  
    <font color='green'>Проделал исследование, получил результат, прекрасно. Чтобы избежать подобного повторения, можно посоветовать технической команде добавлять в выгрузку id клиентов, тогда строки с одинаковыми параметрами не будут являться дубликатами.</font>
    

### Вывод

Вероятно случайные совпадения или повторения строк copy/paste или дублирующейся выгрузкой. Просто уберу из выдачи, их немного

### Лемматизация

In [27]:
from pymystem3 import Mystem
m = Mystem()
from collections import Counter

#Скопировал весь текстовый блок из value_counts, буду его лемматизировать
purpose_text = 'свадьба, на проведение свадьбы, сыграть свадьбу, операции с недвижимостью, покупка коммерческой недвижимости, операции с жильем, покупка жилья для сдачи, операции с коммерческой недвижимостью, покупка жилья, жилье, покупка жилья для семьи, строительство собственной недвижимости, недвижимость, операции со своей недвижимостью, строительство жилой недвижимости, покупка недвижимости, строительство недвижимости, покупка своего жилья, ремонт жилью, покупка жилой недвижимости, на покупку своего автомобиля, заняться высшим образованием, автомобиль, сделка с подержанным автомобилем, свой автомобиль, на покупку подержанного автомобиля, автомобили, на покупку автомобиля, приобретение автомобиля, дополнительное образование, сделка с автомобилем, высшее образование, образование, получение дополнительного образования, получение образования, профильное образование, получение высшего образования, заняться образованием' 

purpose_words = 'образование, свадьба, недвижимость, строительство, жилье, ремонт, автомобиль'.split(', ')


#Лемматизируем таблицу — функция
def lemmatizator(row):
    lemmas = m.lemmatize(row)
    return lemmas

#Лемматизируем таблицу — меняем столбец
df['purpose'] = df['purpose'].apply(lemmatizator) 
df['purpose']

#Функция сранения образцов категорий с леммами в таблице
def compare(row):
    if 'жилье' in row:
        return 'недвижимость'
    for word in purpose_words:
        if word in row:
            return word


#Сравнение образцов категорий с леммами в таблице
df['purpose'] = df['purpose'].apply(compare) 


df['purpose']
df['purpose'].value_counts()


недвижимость    8814
автомобиль      3469
образование     3252
свадьба         1874
Name: purpose, dtype: int64

### Вывод

Так, мне удалось привести столбец purpose к нужному виду — всего лишь 4 категории


<font color='green'>Результат правильный. Только на будущее советую формировать строки типа purpose_text не с помощью копипейста, а в цикле - никогда не знаешь, насколько большой сет данных попадется.</font>

### Категоризация данных

###### Уберем 20 детей — это ошибка и их всего 76

In [28]:
df = df[df.children != 20]


In [29]:
def children_or_not(number):
    if number == 0:
        return 'без детей'
    return 'с детьми'

df['children_category'] = df['children'].apply(children_or_not)
df

children  days_employed  dob_years       education  education_id  \
0             1           8438         42          высшее             0   
1             1           4025         36         среднее             1   
2             0           5623         33         среднее             1   
3             3           4125         32         среднее             1   
4             0            926         27          высшее             0   
...         ...            ...        ...             ...           ...   
17404         1           2351         37  ученая степень             4   
17405         1           4529         43         среднее             1   
17406         1           2113         38         среднее             1   
17407         3           3112         38         среднее             1   
17408         2           1985         40         среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      f   сотрудник     0   
1       женат / замужем                 0      f   сотрудник     0   
2       женат / замужем                 0      m   сотрудник     0   
3       женат / замужем                 0      m   сотрудник     0   
4      гражданский брак                 1      m   компаньон     0   
...                 ...               ...    ...         ...   ...   
17404         в разводе                 3      m   сотрудник     0   
17405  гражданский брак                 1      f   компаньон     0   
17406  гражданский брак                 1      m   сотрудник     1   
17407   женат / замужем                 0      m   сотрудник     1   
17408   женат / замужем                 0      f   сотрудник     0   

       total_income       purpose children_category  
0            253876  недвижимость          с детьми  
1            112080    автомобиль          с детьми  
2            145886  недвижимость         без детей  
3            267629   образование          с детьми  
4            255764  недвижимость         без детей  
...             ...           ...               ...  
17404        115949  недвижимость          с детьми  
17405        224792  недвижимость          с детьми  
17406         89673  недвижимость          с детьми  
17407        244093    автомобиль          с детьми  
17408         82047    автомобиль          с детьми  

[17343 rows x 13 columns]

In [30]:
df.head(50)
children_debt = df.pivot_table(index='children_category', values ='debt', aggfunc =['count', 'mean'])
children_debt.round(2)

count  mean
                    debt  debt
children_category             
без детей          10408  0.08
с детьми            6935  0.09

###### Изучим доход

In [31]:
print('Минимальный доход', df['total_income'].min())

Минимальный доход 21368


In [32]:
print('Максимальный доход', df['total_income'].max())

Максимальный доход 2265604


In [33]:
print('Средний доход', df['total_income'].mean())

Средний доход 174014.89615406792


In [34]:
'Медианный доход', df['total_income'].median()

('Медианный доход', 161386.0)

Разобьем вот так: до 50 тысяч в месяц
от 50 до 100
от 100 до 200
от 200 до 500
от 500 и выше

<font color='green'>Хорошо, что исследовал данные перед назначением категорий.</font>

In [35]:
def income_category(number):
    if number < 50000:
        return 'меньше 50к'
    elif number >= 50000 and number <=100000:
        return 'от 50 до 100'
    elif number > 100000 and number < 200000:
        return 'от 100 до 200'
    elif number >= 200000 and number <=500000:
        return 'от 200 до 500'
    return '500 и выше'

df['income_category'] = df['total_income'].apply(income_category)
df['income_category']

0        от 200 до 500
1        от 100 до 200
2        от 100 до 200
3        от 200 до 500
4        от 200 до 500
             ...      
17404    от 100 до 200
17405    от 200 до 500
17406     от 50 до 100
17407    от 200 до 500
17408     от 50 до 100
Name: income_category, Length: 17343, dtype: object

In [36]:
income_category_debt = df.pivot_table(index='income_category', values ='debt', aggfunc =['count', 'mean'])
income_category_debt

count      mean
                 debt      debt
income_category                
500 и выше        200  0.070000
меньше 50к        191  0.073298
от 100 до 200    9775  0.091151
от 200 до 500    4239  0.072659
от 50 до 100     2938  0.092920

<font color='green'>Все верно, категоризация принимается.  
Но на будущее советую брать цифры для категоризации не наугад, а использовать квантили.</font>

### Вывод

Меньше всего просрочек у семей с пятью детьми — но их всего 9, возможно мало данных. 20 детей у 76 человек — возможно эта ошибка, у них вообще нет просрочек. В целом между отсутствием детей и надичием до 4 включительно разницы нет — просрочку имеет 9 человек из 100


<font color='brown'> 20 детей, определенно это ошибка,как думаешь, можно ли что-то с этим сделать?</font>

Вывел много параметров, может и правда есть 20 детей? Их всего 76, можно удалить или например предположить, что это ошибка ввода — заменить на '2'

### Шаг 3. Ответьте на вопросы

- Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [37]:
children_debt.round(2)


count  mean
                    debt  debt
children_category             
без детей          10408  0.08
с детьми            6935  0.09

### Вывод

20 детей в расчет не берем, считаем ошибкой. Семьи с 5 детьми не имеют просрочек. У заемщика без детей и с тремя детьми просрочка у 8 из 100. Один и двое детей — 9 из 100. Четверо детей — 10 из 100 (1/10)

<font color='red'>Насколько я вижу, разница между отсутствием и наличием детей это 7,5% против 9%. 1,5%, кажется что это немного, но, во-первых, возможно эта цифра интересна заказчику, а во-вторых, если сравнить цифры между собой, то выходит 20% разницы, не так уж и мало?</font>

<br> <font color='blue'>Написал более подробный комментарий без "воды"

<font color='red'>Заметная разница только среди категорий "есть дети" и "нет детей", давай проверим возвращаемость по ним и сформулируем вывод.</font>



###### Просрочка у заемщиков с детьми 9%, без детей — 8%
<font color='green'>Хорошо. Но необязательно для этого было выделять отдельный столбец. Этот вывод можно было сделать и по изначальной сводке с детьми.</font>

- Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [38]:
family_status_debt = df.pivot_table(index='family_status', values ='debt', aggfunc =['count', 'mean'])
family_status_debt.round(2)

count  mean
                        debt  debt
family_status                     
в разводе                954  0.07
вдовец / вдова           398  0.06
гражданский брак        3451  0.10
женат / замужем        10117  0.08
не женат / не замужем   2423  0.11

In [39]:
def married_or_not(row):
    if 'гражданский брак' in row:
        return 'не были женаты/замужем'
    elif 'не женат / не замужем' in row:
        return 'не были женаты/замужем'    
    return 'были женаты/замужем'

df['married_or'] = df['family_status'].apply(married_or_not)
df


children  days_employed  dob_years       education  education_id  \
0             1           8438         42          высшее             0   
1             1           4025         36         среднее             1   
2             0           5623         33         среднее             1   
3             3           4125         32         среднее             1   
4             0            926         27          высшее             0   
...         ...            ...        ...             ...           ...   
17404         1           2351         37  ученая степень             4   
17405         1           4529         43         среднее             1   
17406         1           2113         38         среднее             1   
17407         3           3112         38         среднее             1   
17408         2           1985         40         среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      f   сотрудник     0   
1       женат / замужем                 0      f   сотрудник     0   
2       женат / замужем                 0      m   сотрудник     0   
3       женат / замужем                 0      m   сотрудник     0   
4      гражданский брак                 1      m   компаньон     0   
...                 ...               ...    ...         ...   ...   
17404         в разводе                 3      m   сотрудник     0   
17405  гражданский брак                 1      f   компаньон     0   
17406  гражданский брак                 1      m   сотрудник     1   
17407   женат / замужем                 0      m   сотрудник     1   
17408   женат / замужем                 0      f   сотрудник     0   

       total_income       purpose children_category income_category  \
0            253876  недвижимость          с детьми   от 200 до 500   
1            112080    автомобиль          с детьми   от 100 до 200   
2            145886  недвижимость         без детей   от 100 до 200   
3            267629   образование          с детьми   от 200 до 500   
4            255764  недвижимость         без детей   от 200 до 500   
...             ...           ...               ...             ...   
17404        115949  недвижимость          с детьми   от 100 до 200   
17405        224792  недвижимость          с детьми   от 200 до 500   
17406         89673  недвижимость          с детьми    от 50 до 100   
17407        244093    автомобиль          с детьми   от 200 до 500   
17408         82047    автомобиль          с детьми    от 50 до 100   

                   married_or  
0         были женаты/замужем  
1         были женаты/замужем  
2         были женаты/замужем  
3         были женаты/замужем  
4      не были женаты/замужем  
...                       ...  
17404     были женаты/замужем  
17405  не были женаты/замужем  
17406  не были женаты/замужем  
17407     были женаты/замужем  
17408     были женаты/замужем  

[17343 rows x 15 columns]

In [40]:
family_status_debt2 = df.pivot_table(index='married_or', values ='debt', aggfunc =['count', 'mean'])
family_status_debt2.round(2)

count  mean
                         debt  debt
married_or                         
были женаты/замужем     11469  0.08
не были женаты/замужем   5874  0.10

### Вывод

Просрочку допускают:<br>
вдовец / вдова	— 6%<br>в разводе — 7% <br>женат / замужем — 8% <br>гражданский брак — 10% <br>не женат / не замужем — 11%<br> <br> Надо будет еще построить график зависимости просрочки от возраста — скорее всего не женатые и гражданский брак моложе остальных

<font color='red'> 
    Давай разделим всех на 2 катеогрии: не были женаты (гражданский брак и не женат/не замужем) и были (все остальные), и сделаем по ним вывод.</font>    
    <br> Построил. вывод — те, что не были в браке допускают 10% просрочек, а те, что были — 8%  
    <font color='green'>Правильно замечено, принимается.  
    И еще совет - перенос строки в блоке markdown можно сделать просто двумя пробелами в конце абзаца</font>
    

- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [41]:
income_category_debt.round(2)

count  mean
                 debt  debt
income_category            
500 и выше        200  0.07
меньше 50к        191  0.07
от 100 до 200    9775  0.09
от 200 до 500    4239  0.07
от 50 до 100     2938  0.09

### Вывод

Меньше всего просрочек (6 человек из 100) в категориях дохода до 50к и выше 500.  
От 200 до 500 имеют просрочку 7 человек из 100.  
в остальных категориях 8 из 100

- Как разные цели кредита влияют на его возврат в срок?

In [42]:
purpose_debt = df.pivot_table(index='purpose', values ='debt', aggfunc =['count', 'mean'])
purpose_debt.round(2)

count  mean
              debt  debt
purpose                 
автомобиль    3456  0.10
недвижимость  8784  0.08
образование   3237  0.10
свадьба       1866  0.08

In [48]:
df['total_income'].describe().round()

count      17343.0
mean      174015.0
std       101102.0
min        21368.0
25%       112410.0
50%       161386.0
75%       201944.0
max      2265604.0
Name: total_income, dtype: float64

In [45]:
print(1.124105*10**5)
print(1.613860*10**5)
print(2.019445*10**5)

112410.49999999999
161386.0
201944.50000000003


### Вывод

Меньше всего просрочек на недвижимость и свадьбу — 7%. На образование и автомобиль — 9%

<font color='green'>Выводы по целям и доходам хорошие.</font>

### Шаг 4. Общий вывод

Общий вывод — кредит можно давать всем, максимальный процент просрочки 12%

<font color='red'>Здесь я был бы рад видеть более пространные выводы, результат всей работы, с поставленной задачей, полученными выводами и рекоммендациями для коллег. В конце концов, мы решаем задачу для бизнеса и наша конечная цель - не код, а понятные красивые выводы для заказчика ;).</font>
<br><font color='blue'> В целом просрочка колеблется от 0% до 11%. Рекомендации для коллег — поставить автоматическую проверку данных при вводе.  
    <font color='red'>Я хочу здесь видеть поставленную задачу и 4 пункта с твоими выводами и рекомендациями по каждому вопросу: кому лучше, кому хуже давать кредит в разрезе наших переменных: детей, доходов, семейного положения и целей кредита. Можно ли давать кредит с вероятностью порсрочки 12% пусть решает банк, мы не знаем, выгодно ли ему это, мы стараемся подсказать направление, какие клиенты наиболее безопасны.  
    Без четких выводов и рекомендаций я не смогу принять работу</font>

<font color='green'>Оформление работы хорошее, принимается. Плюс за использование сводных таблиц.</font>
<br><font color='blue'> Оформление доработал с учетом вчерашнего онлайн занятия. Возможно осилю графики еще.  
    <font color='green'>Отлично, я очень рад видеть чистую и аккуратную работу. Как и любой друго человек, который будет читать твой код (а это будет обязательно)</font>
    


## Общий вывод — 2

###### Цели
Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

Рекомендации по заполнению и выгрузке данных — вводить автоматическую проверку полей по категориям, значениям (целым положительным). Ввести автоматическую проверку на дублирующиеся данные

###### Есть ли зависимость между наличием детей и возвратом кредита в срок?
Просрочка у заемщиков с детьми 9%, без детей — 8%

###### Есть ли зависимость между семейным положением и возвратом кредита в срок?
Те, что не были в браке допускают 10% просрочек, а те, что были — 8%

###### Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
Меньше всего просрочек — 6% — в категориях дохода до 50к и выше 500.
От 200 до 500 имеют просрочку 7%. 
В остальных категориях 8%

###### Как разные цели кредита влияют на его возврат в срок?
Меньше всего просрочек на недвижимость и свадьбу — 7%. На образование и автомобиль — 9%

<font color='green'>Вот такие выводы я был бы рад увидеть на месте нашего заказчика.</font>

### Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  открыт файл;
- [x]  файл изучен;
- [x]  определены пропущенные значения;
- [x]  заполнены пропущенные значения;
- [x]  есть пояснение какие пропущенные значения обнаружены;
- [x]  описаны возможные причины появления пропусков в данных;
- [x]  объяснено по какому принципу заполнены пропуски;
- [x]  заменен вещественный тип данных на целочисленный;
- [x]  есть пояснение какой метод используется для изменения типа данных и почему;
- [x]  удалены дубликаты;
- [x]  есть пояснение какой метод используется для поиска и удаления дубликатов;
- [x]  описаны возможные причины появления дубликатов в данных;
- [x]  выделены леммы в значениях столбца с целями получения кредита;
- [x]  описан процесс лемматизации;
- [x]  данные категоризированы;
- [x]  есть объяснение принципа категоризации данных;
- [x]  есть ответ на вопрос "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [x]  есть ответ на вопрос "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [x]  есть ответ на вопрос "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [x]  есть ответ на вопрос "Как разные цели кредита влияют на его возврат в срок?";
- [x]  в каждом этапе есть выводы;
- [x]  есть общий вывод.

<font color='blue'>В целом у тебя хорошая работа, молодец. Хорошо понимаешь все принципы, изучаемые в этом спринте. Но нужно немного доработать.  Больше фокуса на нашу конечную цель - выводы. Еще раз, то, что откомментировано красным, обязательно к исправлению, коричневым - рекомендовано. Жду доработанный проект.</font>

<br><font color='purple'> Спасибо!</font>

<font color='blue'>Ты уверенно движешься к принятию работы, последнее усилие, и ты у цели!</font>

<font color='green'>Подведем итоги. В ходе рыботы ты показал хорошее владение изучаемым материалом:
    умеешь подготавливать, очищать, обогощать данные. Также ты правильно разбил их на категории и проанализировал
    результаты. Заказчику будет что почерпнуть из твоего анализа. Ты учел все замечания и я могу принять работу. Удачи в дальнейшем прохождении курса!</font>